### PR TITLE
Add msgpack codec (based on https://godoc.org/github.com/ugorji/go)

### PR DESCRIPTION
**Pull request by hsanjuan**
My impression so far is that ugorji/go codecs are very mature. This wraps msgpack as multicodec. All very inspired in current mc_cbor.

### CODE REVIEW

**Comment by lgierth** [comment]
`/msgpack` must be added to table.csv in multiformats/multicodec too

**Comment by hsanjuan** [comment]
@lgierth without a code? Just like cbor/bson?

**Comment by lgierth** [comment]
> @lgierth without a code? Just like cbor/bson?

Yes -- multicodec itself doesn't technically need the protocol codes, only multicodec-packed does.

**Comment by hsanjuan** [comment]
Heads up: now it's on the table on the table (https://github.com/multiformats/multicodec/pull/31)

File: msgpack/msgpack.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	// Encode options
+	StructToArray       bool
+	Canonical           bool
+	CheckCircularRef    bool
+	RecursiveEmptyCheck bool
+	Raw                 bool
+	AsSymbols           gocodec.AsSymbolFlag
+
+	// Decode options
+	MapType              reflect.Type
```
Comment (whyrusleeping) [inline msgpack/msgpack.go]: Any reason these are all globals?

File: msgpack/msgpack.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	ErrorIfNoArrayExpand bool
+	SignedInteger        bool
+	MapValueReset        bool
+	InterfaceReset       bool
+	InternString         bool
+	PreferArrayOverSlice bool
+)
+
+func init() {
+	HeaderPath = "/msgpack"
```
Comment (whyrusleeping) [inline msgpack/msgpack.go]: These could be made into constants instead of assigning them in the init here.

File: msgpack/msgpack.go (hunk 1 above)
Comment (whyrusleeping) [inline msgpack/msgpack.go]: I see that they are used as defaults. Perhaps have a `DefaultEncoder` that you can just copy when creating a new decoder/encoder

**Comment by hsanjuan** [comment]
@whyrusleeping yes you're right, I overcomplicated it. Thanks for the good feedback, How does it look now?

**Pull request merged**